cloudencrypt/dual.go

```diff
@@ -0,0 +1,96 @@
+package cloudencrypt
+
+import (
+	"context"
+)
+
+const (
+	mapKeyEncryptedKey   = "key"
+	mapKeyEncryptedValue = "value"
+)
+
+// DualEncryptor implements Encryptor by encrypting the value using the NativeEncryptor
+// and then encrypting the secret key using the provided encryptor.
+type DualEncryptor struct {
+	encryptor Encryptor
+}
+
+func NewDualEncryptor(ctx context.Context, encryptor Encryptor) (*DualEncryptor, error) {
+	return &DualEncryptor{
+		encryptor: encryptor,
+	}, nil
+}
+
+func (encryptor *DualEncryptor) Encrypt(ctx context.Context, value []byte, metadata ...MetadataKV) ([]byte, error) {
+	// Generate a random secret key
+	secretKey, err := generateSecretKey()
+	if err != nil {
+		return nil, err
+	}
+
+	nativeEncryptor, err := NewNativeEncryptor(secretKey)
+	if err != nil {
+		return nil, err
+	}
+
+	defer nativeEncryptor.Close()
+
+	// Encrypt the value using the random secret key
+	encryptedValue, err := nativeEncryptor.Encrypt(ctx, value)
+	if err != nil {
+		return nil, err
+	}
+
+	// Encrypt the secret key
+	encryptedSecretKey, err := encryptor.encryptor.Encrypt(ctx, secretKey, metadata...)
+	if err != nil {
+		return nil, err
+	}
+
+	output := make(map[string]string)
+	output[mapKeyEncryptedKey] = string(encryptedSecretKey)
+	output[mapKeyEncryptedValue] = string(encryptedValue)
+
+	encoded, err := encode(output)
+	if err != nil {
+		return nil, err
+	}
+
+	return encoded, nil
+}
+
+func (encryptor *DualEncryptor) Decrypt(ctx context.Context, encryptedValue []byte, metadata ...MetadataKV) ([]byte, error) {
+	decoded, err := decode(encryptedValue)
+	if err != nil {
+		return nil, err
+	}
+
+	// Decrypt the secret key
+	secretKey, err := encryptor.encryptor.Decrypt(ctx, []byte(decoded[mapKeyEncryptedKey]), metadata...)
+	if err != nil {
+		return nil, err
+	}
+
+	// Decrypt the value using the decrypted secret key
+	nativeEncryptor, err := NewNativeEncryptor(secretKey)
+	if err != nil {
+		return nil, err
+	}
+
+	defer nativeEncryptor.Close()
+
+	decrypted, err := nativeEncryptor.Decrypt(ctx, []byte(decoded[mapKeyEncryptedValue]))
+	if err != nil {
+		return nil, err
+	}
+
+	return decrypted, nil
+}
+
+func (encryptor *DualEncryptor) Close() error {
+	return encryptor.encryptor.Close()
+}
+
+func generateSecretKey() ([]byte, error) {
+	return randomBytes(32)
+}
```

Comment:
This seems like utils function, can be placed in `internal/utils`. I suppose it is not used as library call

Reply:
I don't see what's wrong with this? It's only used here and in tests and it's not exported...

Reply:
Basic library thing. You bother the developers that want to use this library only with exported functions. Utility etc. Goes aside typically to internal module.

Reply:
I don't understand. That's what exporting is for. This function is not exported and won't be used by anyone.

Reply:
https://github.com/golang-standards/project-layout

Reply:
This repository is no standard but highly opinionated trap for Go beginners which I also fell into when I was first learning Go. I categorically refuse using this in any way. See: https://github.com/golang-standards/project-layout/issues/117

Reply:
I don't understand why we have to have discussion about this. I think you just want to proof that you can write code by your taste. Ofc Go compiler leaves you do anything but learning over and over standards for each project is just a trap.
I wish you come with some better solution than saying it is a trap and writing it as you wish so.
This was made when go was 1.14 today we have 10 new releases and noone come up with any new set of golden standards how the project layout should look like.
Can you please draft how it should look like if this is not a reference for our code? I though this is quite a look of our projects.
We currently use `cmd`, `internal`, `pkg`, `build`, `docs`, `scripts`, `third_party`, `tools`, `vendor`, `assets`.
Please come with something else than denial and reference why we should not do something, because I learnt from this thread very little from your side.

Reply:
https://go.dev/doc/modules/layout#package-or-command-with-supporting-packages

Reply:
https://github.com/kubecub/go-project-layout

Reply:
> I don't understand why we have to have discussion about this.

Very simple. I don't blindly follow something unless I'm convinced it's a good thing and I was missing a crucial information about the internal packages. Now I understand the internal packages and moved some code into them. It was very important for me to understand the reasons and benefits rather than just do as I'm told.

https://github.com/golang-standards/project-layout is an immediate red flag because I know how controversial that repository is. That's not to say it doesn't contain some good ideas but it isn't a project to follow blindly - it's highly opinionated layout someone made and published under a misleading name. Many experienced developers disagree with it so any idea coming from that source needs to be verified independently.